File: packages/types/src/events/checkpoint-events.ts
/**
 * Checkpoint Related Event Type Definitions
 */

import type { ID } from "../common.js";
import type { BaseEvent } from "./base.js";

/**
 * Checkpoint creation event type
 */
export interface CheckpointCreatedEvent extends BaseEvent {
  type: "CHECKPOINT_CREATED";
  /** Checkpoint ID */
  checkpointId: ID;
  /** Checkpoint Description */
  description?: string;
}

/**
 * Checkpoint recovery event type
 */
export interface CheckpointRestoredEvent extends BaseEvent {
  type: "CHECKPOINT_RESTORED";
  /** Checkpoint ID */
  checkpointId: ID;
  /** Recovered thread ID */
  threadId: ID;
  /** Checkpoint Description */
  description?: string;
}

/**
 * Checkpoint Deletion Event Type
 */
export interface CheckpointDeletedEvent extends BaseEvent {
  type: "CHECKPOINT_DELETED";
  /** Checkpoint ID */
  checkpointId: ID;
  /** Reason for deletion */
  reason?: "manual" | "cleanup" | "policy";
}

/**
 * Checkpoint failure event type
 */
export interface CheckpointFailedEvent extends BaseEvent {
  type: "CHECKPOINT_FAILED";
  /** Checkpoint ID (if generated) */
  checkpointId?: ID;
  /** Types of failed operations */
  operation: "create" | "restore" | "delete";
  /** error message */
  error: string;
}

File: crates/wf-types/src/events/checkpoint_events.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CheckpointCreatedEvent {
    pub base: super::BaseEvent,
    pub checkpoint_id: super::super::Id,
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CheckpointRestoredEvent {
    pub base: super::BaseEvent,
    pub checkpoint_id: super::super::Id,
    pub execution_id: super::super::Id,
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CheckpointDeletedEvent {
    pub base: super::BaseEvent,
    pub checkpoint_id: super::super::Id,
    pub reason: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
pub struct CheckpointFailedEvent {
    pub base: super::BaseEvent,
    pub checkpoint_id: Option<super::super::Id>,
    pub operation: String,
    pub error: String,
}
